crates/exex/src/manager.rs

```diff
@@ -0,0 +1,396 @@
+    /// The number of events an ExEx has sent to the manager.
+    events_sent: Counter,
+}
+
+/// A handle to an ExEx used by the [`ExExManager`] to communicate with ExEx's.
+///
+/// A handle should be created for each ExEx with a unique ID. The channels returned by
+/// [`ExExHandle::new`] should be given to the ExEx, while the handle itself should be given to the
+/// manager in [`ExExManager::new`].
+#[derive(Debug)]
+pub struct ExExHandle {
+    /// The execution extension's ID.
+    id: String,
+    /// Metrics for an ExEx.
+    metrics: ExExMetrics,
+
+    /// Channel to send [`CanonStateNotification`]s to the ExEx.
+    sender: PollSender<CanonStateNotification>,
+    /// Channel to receive [`ExExEvent`]s from the ExEx.
+    receiver: UnboundedReceiver<ExExEvent>,
+    /// The ID of the next notification to send to this ExEx.
+    next_notification_id: usize,
+
+    /// The finished block number of the ExEx.
+    /// tbd describe how this can change, what None means, what this is used for
+    finished_height: Option<BlockNumber>,
+}
+
+impl ExExHandle {
+    /// Create a new handle for the given ExEx.
+    ///
+    /// Returns the handle, as well as a [`Sender`] for [`ExExEvent`]s and a
+    /// [`Receiver`] for [`CanonStateNotification`]s that should be given to the ExEx.
+    pub fn new(id: String) -> (Self, UnboundedSender<ExExEvent>, Receiver<CanonStateNotification>) {
+        let (canon_tx, canon_rx) = mpsc::channel(1);
+        let (event_tx, event_rx) = mpsc::unbounded_channel();
+
+        (
+            Self {
+                id: id.clone(),
+                metrics: ExExMetrics::new_with_labels(&[("exex", id)]),
+                sender: PollSender::new(canon_tx),
+                receiver: event_rx,
+                next_notification_id: 0,
+                finished_height: None,
+            },
+            event_tx,
+            canon_rx,
+        )
+    }
+
+    /// Reserves a slot in the `PollSender` channel and sends the notification if the slot was
+    /// successfully reserved.
+    ///
+    /// When the notification is sent, it is considered delivered.
+    fn send(
+        &mut self,
+        cx: &mut Context<'_>,
+        (event_id, notification): &(usize, CanonStateNotification),
+    ) -> Poll<Result<(), PollSendError<CanonStateNotification>>> {
+        match self.sender.poll_reserve(cx) {
+            Poll::Ready(Ok(())) => (),
+            other => return other,
+        }
+
+        match self.sender.send_item(notification.clone()) {
+            Ok(()) => {
+                self.next_notification_id = event_id + 1;
+                self.metrics.notifications_processed.increment(1);
+                Poll::Ready(Ok(()))
+            }
+            Err(err) => Poll::Ready(Err(err)),
+        }
+    }
+}
+
+// todo
+#[derive(Metrics)]
+#[metrics(scope = "exex_manager")]
+pub struct ExExManagerMetrics {
+    /// Max size of the internal state notifications buffer.
+    max_capacity: Counter,
```

Comment:
shouldn't it be a Gauge as well?

Reply:
no, this is static, it will not change during runtime

Reply:
yes but counter is always increasing and if we restart Reth with lower `max_capacity` value, it will be weird for Prometheus where the counter will go down